### ISSUE TITLE
Accommodate decryption-from-self in deletion schedule

### ISSUE THREAD

**Issue by bifurcation** [mlswg/mls-protocol#667]
Right now, the deletion schedule requires a key to be consumed if it is used to "encrypt or (successfully) decrypt a message".  This means that a client can't decrypt its own messages, because the key has already been deleted.  On the one hand, this isn't a tragedy, since in principle the client could just hold on to the plaintext.  On the other hand, it requires some lightly painful special-casing for things like proposals and application data, where the message is broadcast.  (On the third hand, receiving Commits from yourself already requires special processing.)

It seems like this could be fixed by just changing the deletion requirement to trigger only on decrypt instead of on encrypt or decrypt.  In scenarios where a sender gets her messages back, the effect would be the same, modulo an RTT.  You would have reduced FS in scenarios where the DS does not echo a message back to a sender.

**Comment by kkohbrok** [comment]
If I understand correctly, the result of this change would be that there is no FS for the sender of any message until the end of the epoch or until the secret tree of that epoch is deleted. So you would always have reduced FS. Just in cases there is no echo, you wouldn't have needed the reduction in the first place.

In any case, I'd prefer keeping the strict deletion schedule at the cost of some slight inconvenience in the implementation. Although as you say, own commits need to be treated differently anyway.

**Comment by bifurcation** [comment]
There wouldn't be FS with regard to the sender of a message until the sender receives the message back.  In the echo case, that is quick; in the no-echo case, it is the end of the epoch + any out-of-order delivery window, as you say.

It depends some on what the application wants as well.  If the application never needs to receive from itself and forgets what it sends after sending it, then the current situation is fine.  If the application remembers what it sends (as with current messengers), then the sender will be holding on to the plaintext anyway, so the notion of FS with regard to the sender is kind of moot.

For completeness, an intermediate position here would be to say:

* MUST delete on successful decrypt
* SHOULD delete on encrypt, unless required by application context (with explanatory text about decrypt-from-self).

That seems worse than the alternatives, though, because it's a hard-to-understand knob for the application to set.

**Comment by Bren2010** [comment]
This would also seem to have a higher likelihood of encryption keys accidentally being reused.

**Comment by bifurcation** [comment]
@Bren2010 - I'm not sure I follow.  Do you mean that because the key would still exist, there's a risk of key management bugs using an old key?

**Comment by Bren2010** [comment]
Yes, I think there's a higher risk of bugs because the key is still around and you're doing additional accounting to separate which ratchet the implementation is using for the next encryption and which is for the next decryption. Being able to delete the key after it was used, regardless of why, is a nice simplification.

The cost of that simplification is that if you want to decrypt-from-self then you need to implement an additional layer that, for example, maps hashes of sent messages to their decrypted/authenticated version. But like people have said, if you're interested in doing decrypt-from-self then you're likely already keeping records of sent messages anyway. So this shouldn't be a big additional burden.

**Comment by tomleavy** [comment]
> An example of note here is the self-deleting messages feature that exist in quite a few messaging apps. Here, you want to send the message and forget it and you explicitly _don't_ want to keep the key material. I think that should be a feature that can be enabled by MLS, either by default or by letting the application choose, i.e. the intermediate position that you describe.

Agreed. I also didn't find the current situation too difficult to implement in practice when I wrote a demo of a websocket relay. It's easy enough to just have to relay filter out the sender from the broadcast or send an ack in place of the message to the sender.
